73 adds nwm v30 to loading utils

samlamont opened the pull request:
* Adds loading functionality for NWM v3.0
* Fixed pydantic configuration models for v2.2 and added for v3.0
* This requires the user to import the correct version of the loading functions based on the date of requested data

mgdenno reviewed (commented):
After initial review, I think we should try to refactor to remove duplicate code across the different NWM version modules.  Also, I should have caught this in a pervious PR, but I think the `models` should be in the models module.

samlamont commented:
- Removed redundant NWM v3.0 code
- Added check for duplicate rows in get_metrics and get_joined_timeseries queries (#69)
- Added control for overwrite file behavior in loading (#77)
- Refactored loading models into a `models/loading` directory

mgdenno commented on tests/test_nwm30_grid_loading.py:
```diff
@@ -0,0 +1,55 @@
+WEIGHTS_FILEPATH = Path(TEST_DIR, "onehuc10_weights.parquet")
+
+
+@pytest.mark.filterwarnings("ignore::RuntimeWarning")
```
@samlamont what's this for?

samlamont replied:
hmm good question.  At one point I was getting a warning related to numpy but it no longer appears.  Maybe I had an issue with my venv.  Will remove these

samlamont replied:
Actually I still see this when running `test_weight_generation.py` using pytest, although it doesn't seem to show up for other tests. Have you come across this?  Googling suggests upgrading numpy, but I'm on the latest version `1.26.2`. Guess I'll just ignore the warning for now

![image](https://github.com/RTIInternational/teehr/assets/7386333/2e4835f0-e187-4ddc-a276-58da855f61d8)

mgdenno commented on src/teehr/loading/usgs/usgs.py:
```diff
@@ -4,7 +4,7 @@
+from teehr.models.loading.utils import ChunkByEnum
 from pydantic import validate_arguments
 
 DATETIME_STR_FMT = "%Y-%m-%dT%H:%M:00+0000"
```
@samlamont we should add overwriting protection to USGS too.

mgdenno commented on tests/test_nwm22_grid_loading.py:
```diff
@@ -14,14 +18,16 @@
                                        TEST_DIR,
                                        False)
 
     json_file = Path(TEST_DIR,
```
@samlamont not sure where to best leave this comment, but in general, I'd say it is probably better to not include test output in the repo.  Have tests write to a folder with a .gitignore that ignores everything.

mgdenno commented on tests/test_nwm30_grid_loading.py:
```diff
@@ -0,0 +1,55 @@
+@pytest.mark.filterwarnings("ignore::RuntimeWarning")
+def test_grid_loading():
+
+    component_paths = [
```
@samlamont should/could thins be two tests?

samlamont replied:
yes good point will do

mgdenno reviewed (changes requested):
@samlamont I left a few comments.  I also made a PR to merge 80 in to 73.  Maybe we can do this before merging 73 to main.

mgdenno commented on tests/test_nwm22_point_loading.py:
```diff
@@ -26,11 +34,17 @@ def test_point_loading():
         True,
```
@samlamont kind of nit picky, but I think adding the argument name before the argument is preferable (num_chunks=100 vs  just 100) esp. when there are lots of arguments.  Otherwise it is hard to tell what is going on.

samlamont commented:
@mgdenno I merged in your PR and addressed comments. Let me know if you see any further issues.  One thing I noticed in `test_nwm20_retro_day()` is that only a single file gets produced (one day) even though the test expects to find two days.  Not sure if `nwm_retro_to_parquet()` is intended to create files for two days here?  Let me know if so and I can take a look at the logic [here](https://github.com/RTIInternational/teehr/blob/bc36dd633798c7072c1114431a1627d9604b248d/src/teehr/loading/nwm_common/retrospective.py#L172)

mgdenno commented on tests/test_nwm22_grid_loading.py:
```diff
@@ -0,0 +1,65 @@
+    assert filecmp.cmp(test_file, json_file, shallow=False)
+
+
+def test_grid_fetch_and_format():
```
@samlamont you think it would make sense to reference the json file in TEST_DIR instead so it does not rely on the results from a previous test to pass?

samlamont replied:
Yes good point, we could move a copy of `nwm.20201218.nwm.t00z.analysis_assim.forcing.tm00.conus.nc.json` into the TEST_DIR and access the file there

mgdenno replied:
couldn't we just use the one we are comparing to instead?

samlamont replied:
well `fetch_and_format` gets parameters from the filename itself, like the day and z-hour of the file to access from GCS https://github.com/RTIInternational/teehr/blob/23daf1fbab6c6c957c1e6110425c53656cdbbf60/src/teehr/loading/nwm_common/grid_utils.py#L96

mgdenno replied:
ah, I see.  i guess you could just use the correct file name since it is in a separate directory from the one where the test will write.

mgdenno commented on tests/test_nwm22_grid_loading.py:
```diff
@@ -0,0 +1,65 @@
+        ignore_missing_file=False
+    )
+
+    json_file = Path(TEMP_DIR,
```
@samlamont would it make sense to reference the response from `build_zarr_references` instead of hard coding?

samlamont replied:
Yes, not sure why I hard-coded the path for `json_file`
